guides-markdown/bcoin-clock.md

```diff
@@ -0,0 +1,560 @@
+BLOCKS_DIR = os.path.expanduser('~') + '/blocks/'
+# global variable to store all the block details
+BLOCKS = {}
+# generic function to load any directory of JSON files into any object passed in
+def readFiles(dict, dir):
+        files = [files for (dirpath, dirnames, files) in os.walk(dir)][0]
+        for index in files:
+                try:
+                        with open (dir + index) as file:
+                                dict[int(index)] = json.load(file)
+                except:
+                        pass
+                sortedDictKeys = sorted(dict)
+                if len(sortedDictKeys) > 20:
+                        del dict[sortedDictKeys[0]]
+
+# run the above function one time to load up the BLOCKS object
+readFiles(BLOCKS, BLOCKS_DIR)
+# Output!
+print json.dumps(BLOCKS, indent=1)
+```
+
+In addition to the node info, address and QR code we got in the last step, this time you should also get a list of 20 blocks and their metadata. We've almost got everything ready for the clock display under the hood! The last bit of internal calculation is to derive the progress of the [difficulty adjustment interval](https://en.bitcoin.it/wiki/Difficulty), and the [block subsidy halving interval.](https://en.bitcoin.it/wiki/Controlled_supply). Since we can get the block height from `getInfo()` it's easy to calculate where we are in each cycle:
+
+```python
+### calculate cycle progress as % and countdown integer given current blockchain height
+def getDiff(height):
+	return {"percent": (height % 2016 / 2016.0) * 100, "countdown": 2016 - (height % 2016)}
+def getHalf(height):
+	return {"percent": (height % 210000 / 210000.0 ) * 100, "countdown": 210000 - (height % 210000)}
+
+info = getInfo()
+height=info['chain']['height']
+print json.dumps(getDiff(height), indent=1)
+print json.dumps(getHalf(height), indent=1)
+```
+
+## Finishing the GUI
+
+What we have built so far is a system that:
+
+1. Launches an SPV Bitcoin node
+
+2. Records metadata about every new block in a file
+
+3. Monitors the status of the node and the blockchain
+
+4. Retrieves Bitcoin addresses on demand from the wallet and generates QR codes
+
+All that remains to package up this structure with a fun graphical interface that refreshes its own status every second *(tick, tick, tick!)*. The final Python script is [available on Github](https://github.com/pinheadmz/bcoin-clock) and you can read through it to see how I added the `curses` Python library to draw ASCII-art to the terminal screen. I added a series of conditional statements to draw the graphics differently depending on the terminal window size. `curses` also allows us to wait for single-key commands from the user and react quickly. I've set up a little menu at the bottom of the screen where a user can zoom the timeline in or out, and also request a deposit address and QR code. You can clone this repository and run the ASCII-art GUI on your own machine with the same methods used in this guide.
```

Comment:
Why not add the final ASCII art code here too? 

The zooming in and out functionality is a really nice touch :D